hoomd/box.py

```diff
@@ -0,0 +1,378 @@
+    This will force Lz=1 and xz=yz=0. init commands that support 2D boxes will
+    pass the dimensionality along to the system. When you assign a new boxdim to
+    an already initialized system, the dimensionality flag is ignored. Changing
+    the number of dimensions during a simulation run is not supported.
+
+    In 2D boxes, *volume* is in units of area.
+
+    .. rubric:: Shorthand notation
+
+    data.boxdim accepts the keyword argument *L=x* as shorthand notation for
+    `Lx=x, Ly=x, Lz=x` in 3D and `Lx=x, Ly=z, Lz=1` in 2D. If you specify both
+    `L=` and `Lx,Ly, or Lz`, then the value for `L` will override the others.
+
+    Examples:
+
+    * Cubic box with given volume: `data.boxdim(volume=V)`
+    * Triclinic box in 2D with given area: `data.boxdim(xy=1.0, dimensions=2,
+      volume=A)`
+    * Rectangular box in 2D with given area and aspect ratio: `data.boxdim(Lx=1,
+      Ly=aspect, dimensions=2, volume=A)`
+    * Cubic box with given length: `data.boxdim(L=10)`
+    * Fully define all box parameters: `data.boxdim(Lx=10, Ly=20, Lz=30, xy=1.0,
+      xz=0.5, yz=0.1)`
+    """
+
+    def __init__(self, Lx=1.0, Ly=1.0, Lz=1.0, xy=0.0, xz=0.0, yz=0.0):
+        self._cpp_obj = _hoomd.BoxDim(Lx, Ly, Lz)
+        self._cpp_obj.setTiltFactors(xy, xz, yz)
+        self._lattice_vectors = LatticeVectors(self._cpp_obj)
+
+    # Dimension based properties
+    @property
+    def dimension(self):
+        return 2 if self.Lz == 0 else 3
+
+    @property
+    def is2D(self):
+        return self.Lz == 0
+
+    # Length based properties
+    @property
+    def L(self):
+        return to_three_array(self._cpp_obj.getL())
+
+    @L.setter
+    def L(self, new_L):
+        if len(new_L) == 2:
+            new_L = [new_L[0], new_L[1], 0]
+        self._cpp_obj.setL(make_scalar3(new_L))
+
+    @property
+    def Lx(self):
+        return self.L[0]
+
+    @Lx.setter
+    def Lx(self, value):
+        L = self._cpp_obj.getL()
+        L.x = float(value)
+        self.L = L
+
+    @property
+    def Ly(self):
+        return self.L[1]
+
+    @Ly.setter
+    def Ly(self, value):
+        L = self._cpp_obj.getL()
+        L.y = float(value)
+        self.L = L
+
+    @property
+    def Lz(self):
+        return self.L[2]
+
+    @Lz.setter
+    def Lz(self, value):
+        L = self._cpp_obj.getL()
+        L.z = float(value)
+        self.L = L
+
+    # Box tilt based properties
+    @property
+    def tilts(self):
+        return np.array([self.xy, self.xz, self.yz])
+
+    @tilts.setter
+    def tilts(self, new_tilts):
+        if isinstance(new_tilts, (float, int)) or len(new_tilts) == 1:
+            if self.is2D:
+                self._cpp_obj.setTiltFactors(new_tilts[0], 0, 0)
+            else:
+                raise ValueError("Must specify 3 tilt factors for 3D box.")
+        elif len(new_tilts) == 3:
+            self._cpp_obj.setTiltFactors(new_tilts[0], new_tilts[1],
```

Comment:
I don't know if it is in scope to change the underlying C++ class here since it is used so ubiquitelessly throughout HOOMD. I can do a quick grep and see, though. I could also just add methods to set each individual `L`.

Reply:
One can safely add methods to BoxDim C++. Just don't change the behavior or remove any existing methods at this time. That could be done in a targeted refactoring branch after the API work is complete.

Reply:
Let's table this for now, for a potential refactoring of the `BoxDim` class itself. That would not change the Python API.